### PR TITLE
Improve metadata lock granularity

### PR DESCRIPTION
**Pull request by arutk**

### CODE REVIEW

File: src/engine/engine_pt.c [hunk 1]
```diff
@@ -131,13 +128,16 @@ int ocf_read_pt(struct ocf_request *req)
...
-		} else {
-			/* No mapped cache lines, no need to get lock */
-			lock = OCF_LOCK_ACQUIRED;
 		}
 	}
 
-	OCF_METADATA_UNLOCK_RD(); /*- END Metadata RD access -----------------*/
+	if (lock != OCF_LOCK_ACQUIRED) {
+		ocf_req_hash_lock_upgrade(req);
+		lock =ocf_req_async_lock_rd(req, ocf_engine_on_resume);
```
Comment (robertbaldyga) [inline src/engine/engine_pt.c]: Missing space.

File: src/concurrency/ocf_metadata_concurrency.c [hunk 2]
```diff
@@ -4,22 +4,249 @@
...
+	ocf_metadata_hash_unlock(metadata_lock, hash, OCF_METADATA_WR);
+	ocf_metadata_end_shared_access(metadata_lock);
+}
+
+/* Iterate over hash buckets for all core lines in the request. Each hash bucket
+ * is visited only once. The order in which iterator visits hash buckets is
+ * strict - for all requests with core lines belonging to hash buckets no A and
+ * B, the iterator always visits either A or B first.
+ */
+#define for_each_req_hash(req, entries, i, pos) \
```
[the remark this answers is not in the document]
Comment (arutk) [reply src/concurrency/ocf_metadata_concurrency.c]: good idea, thanks

File: src/engine/engine_zero.c [hunk 3]
```diff
@@ -143,6 +140,9 @@ void ocf_engine_zero_line(struct ocf_request *req)
 
 	ENV_BUG_ON(req->core_line_count != 1);
 
+	/* No hash bucket locking - this function must be called under global
```
Comment (imjfckm) [inline src/engine/engine_zero.c]: To what this comment applies? It isn't above any fcn call and it's not in fcn header.

Comment (arutk) [reply src/engine/engine_zero.c]: It applies to ocf_engine_zero_line. Typically all other engines lock hash buckets around this line. This very engine is unique, I wanted to explain this here. I have edited this comment to make this clear - let me know what you think.

**Comment by arutk** [comment]
retest this please

File: src/engine/engine_zero.c (hunk 3 above)
Comment (imjfckm) [reply src/engine/engine_zero.c]: Yup, now it's clearer!

**Pull request merged**